### PR TITLE
Fix VPI timed callbacks

### PR DESCRIPTION
**Pull request by mtdudek**
This PR fixes timed callbacks. So to not break compatibility with older software,
timed callbacks still must be called with callCbs method. NextDeadline will now return lowest of all available timed callbacks.

Signed-off-by: Maciej Dudek <mdudek@antmicro.com>

### CODE REVIEW

File: include/verilated_vpi.cpp [hunk 1]
```diff
@@ -565,8 +585,14 @@ class VerilatedVpiImp final {
     }
     static void cbTimedRemove(uint64_t id, QData time) {
         // Id might no longer exist, if already removed due to call after event, or teardown
-        const auto it = s().m_timedCbs.find(std::make_pair(time, id));
+        auto it = s().m_timedCbs.find(std::make_pair(time, id));
         if (VL_LIKELY(it != s().m_timedCbs.end())) it->second.invalidate();
+        it = s().m_readWriteCbs.find(std::make_pair(time, id));
+        if (VL_LIKELY(it != s().m_readWriteCbs.end())) it->second.invalidate();
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:591]: Remove VL_LIKELY in whole function given we no longer have a single list we would expect it to be in.

File: include/verilated_vpi.cpp [hunk 2]
```diff
@@ -589,11 +615,50 @@ class VerilatedVpiImp final {
...
+        const QData time = VL_TIME_Q();
+        std::map<std::pair<QData, uint64_t>, VerilatedVpiCbHolder>* vec;
+        bool called = false;
+        if (reason == cbNextSimTime) {
+            vec = &s().m_nextSimTimeCbs;
+        } else if (reason == cbReadWriteSynch) {
+            vec = &s().m_readWriteCbs;
+        } else {
+            vec = &s().m_readOnlyCbs;
+        }
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:639]: Should we instead check for cbReadOnly and assert(0) if not?  (I understand upper code will only call in these cases, but we might change that someday.)

File: include/verilated_vpi.cpp [hunk 3]
```diff
@@ -589,11 +615,50 @@ class VerilatedVpiImp final {
...
+                const auto last_it = it;
+                ++it;
+                if (VL_UNLIKELY(!ho.invalid())) {
+                    VL_DEBUG_IF_PLI(
+                        VL_DBG_MSGF("- vpi: timed_callback id=%" PRId64 "\n", ho.id()););
+                    ho.invalidate();
+                    (ho.cb_rtnp())(ho.cb_datap());
+                    called = true;
+                }
+                vec->erase(last_it);
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:652]: Erase is very slow on a vector, can we use a different data structure?

File: include/verilated_vpi.cpp [hunk 4]
```diff
@@ -1315,9 +1382,13 @@ vpiHandle vpi_register_cb(p_cb_data cb_data_p) {
         VerilatedVpiImp::cbTimedAdd(id, cb_data_p, abstime);
         return vop->castVpiHandle();
     }
-    case cbReadWriteSynch:  // FALLTHRU // Supported via vlt_main.cpp
-    case cbReadOnlySynch:  // FALLTHRU // Supported via vlt_main.cpp
-    case cbNextSimTime:  // FALLTHRU // Supported via vlt_main.cpp
+    case cbNextSimTime: {
+        const QData abstime = VL_TIME_Q() + 1;
+        const uint64_t id = VerilatedVpiImp::nextCallbackId();
+        VerilatedVpioTimedCb* const vop = new VerilatedVpioTimedCb{id, abstime};
```
Comment (wsnyder) [inline include/verilated_vpi.cpp:1388]:
```suggestion
        VerilatedVpioTimedCb* const vop = new VerilatedVpioTimedCb{id, abstime};
```

**Review by wsnyder — COMMENTED** [review]
Did you check the modified test passes on another simulator (ideally commercial)?

File: include/verilated.h [hunk 5]
```diff
@@ -481,7 +487,10 @@ class VerilatedContext VL_NOT_FINAL {
     /// Set current simulation time. See time() for side effect details
     void time(uint64_t value) VL_MT_SAFE { m_s.m_time = value; }
     /// Advance current simulation time. See time() for side effect details
-    void timeInc(uint64_t add) VL_MT_UNSAFE { m_s.m_time += add; }
+    void timeInc(uint64_t add) VL_MT_UNSAFE {
+        m_s.m_time += add;
+        if (add) m_s.m_gotEvalInTime = true;
```
[the remark this answers is not in the document]
Comment (mtdudek) [reply include/verilated.h:492]: True, that was, at the beginning, aimed at adding checks for registering cbAtStartOfSimTime with time value of 0. It is allowed only if we are before any evaluation. Later I've didn't add cbAtStartOfSimTime callback to list of supported ones. So I'll remove this dead code. (May be useful in future).

File: include/verilated_vpi.cpp [hunk 6]
```diff
@@ -520,7 +520,8 @@ class VerilatedVpiImp final {
 
     // All only medium-speed, so use singleton function
     VpioCbList m_cbObjLists[CB_ENUM_MAX_VALUE];  // Callbacks for each supported reason
-    VpioTimedCbs m_timedCbs;  // Time based callbacks
+    VpioTimedCbs m_timedCbs, m_readWriteCbs, m_readOnlyCbs,
+        m_nextSimTimeCbs;  // Time based callbacks
```
[the remark this answers is not in the document]
Comment (mtdudek) [reply include/verilated_vpi.cpp:524]: I'll change this as suggested

File: include/verilated_vpi.cpp (hunk 4 above)
Comment (mtdudek) [reply include/verilated_vpi.cpp:1388]: Sorry but I don't know what you're suggesting here. Was your intention to remove 
` const QData abstime = VL_TIME_Q() + 1;`  and use
` VerilatedVpioTimedCb* const vop = new VerilatedVpioTimedCb{id, VL_TIME_Q() + 1};`

Comment (wsnyder) [reply include/verilated_vpi.cpp:1388]: The change is just to use C++ {} instead of () in constructors in new code.

**Comment by wsnyder** [comment]
Do you think you will be getting back to this?

**Comment by mtdudek** [comment]
Resently I had no time to work on it, but I'll get back to it.

**Comment by wsnyder** [comment]
#2778 was fixed recently,thanks fir adding that.

**Pull request closed without merging**